### PR TITLE
Fix workflows for slash commands /format and /test-gmt-dev

### LINKED ISSUE #1816: Format and test-gmt-dev slash commands not working properly
**Description of the problem**

The `/format` and `/test-gmt-dev` slash commands added in #646 and #831 respectively seem to have stopped working like they used to, and probably for a while now to be honest. I think GitHub made some changes in regard to bot permissions some time ago (though I can't pin an exact date), and a lot of these tools with auto-commit functionality just broke.

To be specific:
1. For the `/format` command, an action is triggered and runs to completion, but no formatting commit actually happens. E.g.:
  - https://github.com/GenericMappingTools/pygmt/pull/1719#issuecomment-1065125258
  - https://github.com/GenericMappingTools/pygmt/pull/1812#issuecomment-1066258411
2. The `/test-gmt-dev` command, I think it does run (?), but you have to jump to the 'Actions' tab to see it run which is very inconvenient. From memory, I think there was another problem with how it wasn't testing the changes in the PR branch properly, but some rebased commit instead? E.g.:
  - https://github.com/GenericMappingTools/pygmt/pull/1577#issuecomment-944877093
  - https://github.com/GenericMappingTools/pygmt/pull/517#issuecomment-777194521

So, we'll either need to update these convenience slash commands, or remove them.

### PR DESCRIPTION
**Pull request by seisman**
**Description of proposed changes**

As reported in #1816, slash commands `/format` and `/test-gmt-dev` no longer works as expected.

The main cause is that the two workflows always check out the main branch because `ref : ${{ github.event.pull_request.head.sha }}` is empty for a `repository_dispatch` workflow. 

This PR fixes the issue by changing the `ref` to `${{ github.event.client_payload.pull_request.head.ref }}`.

Address #1816.

---

It's impossible to test this PR without merging it because `repository_dispatch` triggered by slash commands can only run workflows in the main branch. 

To test changes in this PR, I have made [the same changes to the main branch of my fork](https://github.com/seisman/pygmt/commit/6c47f7dc69135b4b8a544f2bbdef797656d25ef0) (actually more changes are made because my fork can't use `tibdex/github-app-token`). A sample PR is available at https://github.com/seisman/pygmt/pull/2. As you can see, both commands now work (see https://github.com/seisman/pygmt/actions/runs/2033270957 and https://github.com/seisman/pygmt/actions/runs/2033287345).

### CODE REVIEW

File: .github/workflows/ci_tests_dev.yaml
Comment (weiji14) [reply .github/workflows/ci_tests_dev.yaml:70]: So the slash command at https://github.com/seisman/pygmt/pull/2#issuecomment-1077399124 triggers a GMT Dev Test build, but it still won't show up under the 'Checks' tab? I.e. we still need to navigate to the 'Actions' tab and find it?

**Review by weiji14 — APPROVED** [review]
Test failures seem unrelated (but a bit worrying, need to see what's going on). Ok to merge after second approval.

**Comment by meghanrjones** [comment]
> Test failures seem unrelated (but a bit worrying, need to see what's going on).

The ubuntu/macOS failures are explained by https://github.com/GenericMappingTools/pygmt/issues/1851. The windows tests are more concerning. I could take a look tomorrow.

**Comment by seisman** [comment]
> > Test failures seem unrelated (but a bit worrying, need to see what's going on).
> 
> The ubuntu/macOS failures are explained by #1851. The windows tests are more concerning. I could take a look tomorrow.

The Windows failures are likely related to the recent GDAL rebuild. See https://github.com/conda-forge/gdal-feedstock/issues/616 for the issue report.

**Pull request merged**
